### ISSUE TITLE
The nms seems not working?

### ISSUE THREAD

**Issue by CPFLAME** [yhenon/pytorch-retinanet#27]
Thanks for your nice code.
In visualize.py I changed the code:
`retinanet = model.resnet50(num_classes=dataset_val.num_classes(),)
retinanet.load_state_dict(torch.load(parser.model))`
And I ran this command：
`python visualize.py --dataset coco --coco_path /data/COCO --model models/coco_resnet_50_map_0_335_state_dict.pt`
but I got this result:
![image](https://user-images.githubusercontent.com/13843079/45668928-fced7b80-bb50-11e8-8567-1a55a9a7e457.png)

**Comment by yhenon** [comment]
Hi, 
I'm looking into this issue. Could you please indicate:
- Were you able to compile the NMS extension without any errors?
- Is this running in a jupyer notebook ?

Thanks.

**Comment by CPFLAME** [comment]
Thanks for your reply.
1. I think I compile the NMS extension without any errors: 
`root@9a42af0f41ef:/pytorch-retinanet/lib# sh build.sh 
Compiling nms kernels by nvcc...
Including CUDA code.
/pytorch-retinanet/lib/nms
generating /tmp/tmpm1ZyBj/_nms.c
setting the current directory to '/tmp/tmpm1ZyBj'
running build_ext
building '_nms' extension
creating pytorch-retinanet
creating pytorch-retinanet/lib
creating pytorch-retinanet/lib/nms
creating pytorch-retinanet/lib/nms/src
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fno-strict-aliasing -Wdate-time -D_FORTI
FY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -DWITH_CUDA -I/usr/local/lib/python2.7/dist-packages/torch/utils/ffi/../../lib/include -I/usr/local/lib/python2.7/dist-packages/torch/utils/ffi/../../lib/include/TH -I/usr/local/lib/python2.7/dist-packages/torch/utils/ffi/../../lib/include/THC -I/usr/local/cuda/include -I/usr/include/python2.7 -c _nms.c -o ./_nms.o -std=c99 -std=c99x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fno-strict-aliasing -Wdate-time -D_FORTI
FY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -DWITH_CUDA -I/usr/local/lib/python2.7/dist-packages/torch/utils/ffi/../../lib/include -I/usr/local/lib/python2.7/dist-packages/torch/utils/ffi/../../lib/include/TH -I/usr/local/lib/python2.7/dist-packages/torch/utils/ffi/../../lib/include/THC -I/usr/local/cuda/include -I/usr/include/python2.7 -c /pytorch-retinanet/lib/nms/src/nms.c -o ./pytorch-retinanet/lib/nms/src/nms.o -std=c99 -std=c99x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fno-strict-aliasing -Wdate-time -D_FORTI
FY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -DWITH_CUDA -I/usr/local/lib/python2.7/dist-packages/torch/utils/ffi/../../lib/include -I/usr/local/lib/python2.7/dist-packages/torch/utils/ffi/../../lib/include/TH -I/usr/local/lib/python2.7/dist-packages/torch/utils/ffi/../../lib/include/THC -I/usr/local/cuda/include -I/usr/include/python2.7 -c /pytorch-retinanet/lib/nms/src/nms_cuda.c -o ./pytorch-retinanet/lib/nms/src/nms_cuda.o -std=c99 -std=c99/pytorch-retinanet/lib/nms/src/nms_cuda.c: In function ‘gpu_nms’:
/pytorch-retinanet/lib/nms/src/nms_cuda.c:29:35: warning: initialization from incompatible pointer type [-Wincompatible-p
ointer-types]   unsigned long long* mask_flat = THCudaLongTensor_data(state, mask);
                                   ^
/pytorch-retinanet/lib/nms/src/nms_cuda.c:37:40: warning: initialization from incompatible pointer type [-Wincompatible-p
ointer-types]   unsigned long long * mask_cpu_flat = THLongTensor_data(mask_cpu);
                                        ^
/pytorch-retinanet/lib/nms/src/nms_cuda.c:40:39: warning: initialization from incompatible pointer type [-Wincompatible-p
ointer-types]   unsigned long long* remv_cpu_flat = THLongTensor_data(remv_cpu);
                                       ^
/pytorch-retinanet/lib/nms/src/nms_cuda.c:23:7: warning: unused variable ‘boxes_dim’ [-Wunused-variable]
   int boxes_dim = THCudaTensor_size(state, boxes, 1);
       ^
x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -fno-strict-
aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wl,-Bsymbolic-functions -Wl,-z,relro -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security ./_nms.o ./pytorch-retinanet/lib/nms/src/nms.o ./pytorch-retinanet/lib/nms/src/nms_cuda.o /pytorch-retinanet/lib/nms/src/cuda/nms_kernel.cu.o -o ./_nms.so`

2. I run it in docker(on a Linux machine) instread of jupyter notebook.
Thanks.

**Comment by qq184861643** [comment]
@yhenon Hi have you figured out the problem of nms yet?I think I confront the same problem.Thx

**Comment by mahaling** [comment]
Here's a bug I found in compiling nms.. hope it is useful to someone.
The build.sh file is saying "$CUDA_ARCH" as a parameter to nvcc command which was not set inside my docker environment. 
The actual nvcc command goes like the one in the link below. 
https://github.com/multimodallearning/pytorch-mask-rcnn

I replaced $CUDA_ARCH with --arch=sm_52 (I have a titanx).

I was getting a "no gpu_nms" module found error before and it was gone after compiling nms properly.

Hope this helps someone.

**Comment by foralliance** [comment]
@qq184861643  HI
Agree with u.

In my experiment, although **visualize** works fine, there was no problem. But i think we should **do nms once for each category when using NMS**.

@yhenon 
hope your help!!
many many thanks!!

**Comment by Lam1360** [comment]
I come across the same problem! The NMS extension has been compiled without any errors. I don't know how to solve it. Can anyone help me?

**Comment by Lam1360** [comment]
Hey, I solve this problem! 
The problem lies in lib/build.sh. What I do is just to add '-gencode arch=compute_70,code=sm_70' (according to my GPU device RTX2070) to the CUDA_ARCH and recompile the NMS extension. Now everything works well! 
Hope this help someone and sorry for my poor English.

**Comment by allenyh** [comment]
> Hey, I solve this problem!
> The problem lies in lib/build.sh. What I do is just to add '-gencode arch=compute_70,code=sm_70' (according to my GPU device RTX2070) to the CUDA_ARCH and recompile the NMS extension. Now everything works well!
> Hope this help someone and sorry for my poor English.

Lam, may I ask for the version of CUDA and Pytorch of your machine?
